src/pip/_internal/utils/pyc_compile.py

```diff
@@ -0,0 +1,170 @@
+# -------------------------------------------------------------------------- #
+# NOTE: Importing from pip's internals or vendored modules should be AVOIDED
+#       so this module remains fast to import, minimizing the overhead of
+#       spawning a new bytecode compiler worker.
+# -------------------------------------------------------------------------- #
+
+import compileall
+import importlib
+import os
+import sys
+import warnings
+from collections.abc import Callable, Iterable, Iterator
+from contextlib import contextmanager, redirect_stdout
+from io import StringIO
+from pathlib import Path
+from typing import TYPE_CHECKING, Literal, NamedTuple, Optional, Protocol, Union
+
+if TYPE_CHECKING:
+    from pip._vendor.typing_extensions import Self
+
+WorkerSetting = Union[int, Literal["auto"]]
+
+CODE_SIZE_THRESHOLD = 1000 * 1000  # 1 MB of .py code
+WORKER_LIMIT = 8
+
+
+@contextmanager
+def _patch_main_module_hack() -> Iterator[None]:
+    """Temporarily replace __main__ to reduce the worker startup overhead.
+
+    concurrent.futures imports the main module while initializing new workers
+    so any global state is retained in the workers. Unfortunately, when pip
+    is run from a console script wrapper, the wrapper unconditionally imports
+    pip._internal.cli.main and everything else it requires. This is *slow*.
+
+    The compilation code does not depend on any global state, thus the costly
+    re-import of pip can be avoided by replacing __main__ with any random
+    module that does nothing.
+    """
+    original_main = sys.modules["__main__"]
+    sys.modules["__main__"] = sys.modules["pip"]
+    try:
+        yield
+    finally:
+        sys.modules["__main__"] = original_main
+
+
+class CompileResult(NamedTuple):
+    py_path: str
+    pyc_path: str
+    is_success: bool
+    compile_output: str
+
+
+def _compile_single(py_path: Union[str, Path]) -> CompileResult:
+    # compile_file() returns True silently even if the source file is nonexistent.
+    if not os.path.exists(py_path):
+        raise FileNotFoundError(f"Python file '{py_path!s}' does not exist")
+
+    with warnings.catch_warnings(), redirect_stdout(StringIO()) as stdout:
+        warnings.filterwarnings("ignore")
+        success = compileall.compile_file(py_path, force=True, quiet=True)
+    pyc_path = importlib.util.cache_from_source(py_path)  # type: ignore[arg-type]
+    return CompileResult(
+        str(py_path), pyc_path, success, stdout.getvalue()  # type: ignore[arg-type]
+    )
```

Comment:
These type ignores are due to incorrect type annotations in typeshed. I've filed PRs to fix them, but they haven't been incorporated in a mypy release yet.

- https://github.com/python/typeshed/pull/13509
- https://github.com/python/typeshed/pull/13504